oak/server/grpc_client_node.cc

```diff
@@ -0,0 +1,193 @@
+    : NodeThread(runtime, name),
+      channel_(grpc::CreateChannel(grpc_address, grpc::InsecureChannelCredentials())),
+      stub_(new grpc::GenericStub(channel_)) {
+  LOG(INFO) << "Created gRPC client node for " << grpc_address;
+}
+
+bool GrpcClientNode::HandleInvocation(MessageChannelReadHalf* invocation_channel) {
+  // Expect to receive a pair of handles:
+  //  - Handle to the read half of a channel that holds the request, serialized
+  //    as a GrpcRequest.
+  //  - Handle to the write half of a channel to send responses down, each
+  //    serialized as a GrpcResponse.
+  ReadResult invocation = invocation_channel->Read(INT_MAX, INT_MAX);
+  if (invocation.required_size > 0) {
+    LOG(ERROR) << "Message size too large: " << invocation.required_size;
+    return false;
+  }
+  if (invocation.msg->data.size() != 0) {
+    LOG(ERROR) << "Unexpectedly received data in invocation";
+    return false;
+  }
+  if (invocation.msg->channels.size() != 2) {
+    LOG(ERROR) << "Wrong number of channels " << invocation.msg->channels.size()
+               << " in invocation";
+    return false;
+  }
+
+  std::unique_ptr<ChannelHalf> half0 = std::move(invocation.msg->channels[0]);
+  auto channel0 = absl::get_if<std::unique_ptr<MessageChannelReadHalf>>(half0.get());
+  if (channel0 == nullptr) {
+    LOG(ERROR) << "First channel accompanying invocation is write-direction";
+    return false;
+  }
+  std::unique_ptr<MessageChannelReadHalf> req_channel = std::move(*channel0);
+
+  std::unique_ptr<ChannelHalf> half1 = std::move(invocation.msg->channels[1]);
+  auto channel1 = absl::get_if<std::unique_ptr<MessageChannelWriteHalf>>(half1.get());
+  if (channel1 == nullptr) {
+    LOG(ERROR) << "Second channel accompanying invocation is read-direction";
+    return false;
+  }
+  std::unique_ptr<MessageChannelWriteHalf> rsp_channel = std::move(*channel1);
+
+  // Expect to read a single request out of the request channel.
+  // TODO(#97): support client-side streaming
+  ReadResult req_result = req_channel->Read(INT_MAX, INT_MAX);
+  if (req_result.required_size > 0) {
+    LOG(ERROR) << "Message size too large: " << req_result.required_size;
+    return false;
+  }
+  if (req_result.msg->channels.size() != 0) {
+    LOG(ERROR) << "Unexpectedly received channel handles in request channel";
+    return false;
+  }
+  GrpcRequest grpc_req;
+  grpc_req.ParseFromString(std::string(req_result.msg->data.data(), req_result.msg->data.size()));
+  std::string method_name = grpc_req.method_name();
+  const grpc::string& req_data = grpc_req.req_msg().value();
+
+  // Use a completion queue together with a generic client reader/writer to
+  // perform the method invocation.
+  LOG(INFO) << "invoke method " << method_name;
+  grpc::ClientContext ctx;
+  grpc::CompletionQueue cq;
+  std::unique_ptr<grpc::GenericClientAsyncReaderWriter> call =
+      stub_->PrepareCall(&ctx, method_name, &cq);
+  call->StartCall(tag(1));
```

Comment:
Are the values of these tags important? Do they need to be kept in sync with anywhere else?

Reply:
No, they're not really used here because the client is single-method and synchronous.  Added a comment.

Reply:
No, they're not really used here because the client is single-method and synchronous.  Added a comment.